File: src/types/llm.ts
import { BaseEntity, EntityStatus } from './common';

// LLM types
export type HostingEnvironment = 'azure_ai_foundry' | 'aws_bedrock' | 'aws_sagemaker' | 'gcp_vertex_ai' | 'custom_deployment';
export type GCPAuthMethod = 'adc' | 'service_account' | 'api_key';
export type GCPModelType = 'foundation' | 'fine_tuned' | 'custom';
export type CustomDeploymentLocation = 'cloud' | 'on_premise' | 'hybrid';
export type CustomAPICompatibility = 'openai_compatible' | 'anthropic_compatible' | 'custom';
export type CustomAuthMethod = 'api_key_header' | 'bearer_token' | 'basic_auth' | 'oauth2' | 'none';

export interface LLM extends BaseEntity {
  name: string;
  description?: string;
  model_name: string;
  hosting_environment: HostingEnvironment;
  
  // Azure AI Foundry fields
  azure_endpoint_url?: string;
  azure_api_key?: string;
  azure_deployment_name?: string;
  
  // AWS Bedrock fields
  aws_region?: string;
  aws_access_key_id?: string;
  aws_secret_access_key?: string;
  aws_model_id?: string;
  
  // AWS SageMaker fields
  aws_sagemaker_endpoint_name?: string;
  aws_content_handler_class?: string;
  
  // Google Cloud Vertex AI fields
  gcp_project_id?: string;
  gcp_region?: string;
  gcp_auth_method?: GCPAuthMethod;
  gcp_service_account_key?: string;
  gcp_api_key?: string;
  gcp_model_type?: GCPModelType;
  gcp_model_name?: string;
  
  // Custom Deployment fields
  custom_deployment_location?: CustomDeploymentLocation;
  custom_llm_provider?: string;
  custom_api_endpoint_url?: string;
  custom_api_compatibility?: CustomAPICompatibility;
  custom_auth_method?: CustomAuthMethod;
  custom_auth_header_name?: string;
  custom_auth_key_prefix?: string;
  custom_auth_api_key?: string;
  custom_auth_username?: string;
  custom_auth_password?: string;
  custom_oauth2_token_url?: string;
  custom_oauth2_client_id?: string;
  custom_oauth2_client_secret?: string;
  
  // Model configuration
  temperature?: number;
  max_tokens?: number;
  top_p?: number;
  top_k?: number;
  frequency_penalty?: number;
  presence_penalty?: number;
  stop_sequences?: string;
  
  // System fields
  enabled: boolean;
  status: EntityStatus;
  usage_stats?: LLMUsageStats;
}

export interface LLMUsageStats {
  requests_today: number;
  tokens_used: number;
  rate_limit: number;
}

export interface CreateLLMRequest {
  name: string;
  description?: string;
  model_name: string;
  hosting_environment: HostingEnvironment;
  
  // Azure AI Foundry fields
  azure_endpoint_url?: string;
  azure_api_key?: string;
  azure_deployment_name?: string;
  
  // AWS Bedrock fields
  aws_region?: string;
  aws_access_key_id?: string;
  aws_secret_access_key?: string;
  aws_model_id?: string;
  
  // AWS SageMaker fields
  aws_sagemaker_endpoint_name?: string;
  aws_content_handler_class?: string;
  
  // Google Cloud Vertex AI fields
  gcp_project_id?: string;
  gcp_region?: string;
  gcp_auth_method?: GCPAuthMethod;
  gcp_service_account_key?: string;
  gcp_api_key?: string;
  gcp_model_type?: GCPModelType;
  gcp_model_name?: string;
  
  // Custom Deployment fields
  custom_deployment_location?: CustomDeploymentLocation;
  custom_llm_provider?: string;
  custom_api_endpoint_url?: string;
  custom_api_compatibility?: CustomAPICompatibility;
  custom_auth_method?: CustomAuthMethod;
  custom_auth_header_name?: string;
  custom_auth_key_prefix?: string;
  custom_auth_api_key?: string;
  custom_auth_username?: string;
  custom_auth_password?: string;
  custom_oauth2_token_url?: string;
  custom_oauth2_client_id?: string;
  custom_oauth2_client_secret?: string;
  
  // Model configuration
  temperature?: number;
  max_tokens?: number;
  top_p?: number;
  top_k?: number;
  frequency_penalty?: number;
  presence_penalty?: number;
  stop_sequences?: string;
  
  // System fields
  enabled: boolean;
  status: EntityStatus;
}

export interface UpdateLLMRequest extends Partial<CreateLLMRequest> {}

export interface LLMModel {
  id: string;
  name: string;
  description: string;
}

File: portal-be/app/models/llm.py
"""
LLM model
"""
from sqlalchemy import Column, String, Boolean, Text, JSON, Integer, Float
from .base import BaseModel


class LLM(BaseModel):
    __tablename__ = "llms"

    name = Column(String(255), nullable=False)
    description = Column(Text)
    
    # Hosting Environment
    hosting_environment = Column(String(100), nullable=False)  # azure_ai_foundry, aws_bedrock, aws_sagemaker, gcp_vertex_ai, custom_deployment
    
    # Azure AI Foundry fields
    azure_endpoint_url = Column(String(500))
    azure_api_key = Column(Text)  # Encrypted
    azure_deployment_name = Column(String(255))
    
    # AWS fields (Bedrock & SageMaker)
    aws_region = Column(String(50))
    aws_access_key_id = Column(Text)  # Encrypted
    aws_secret_access_key = Column(Text)  # Encrypted
    aws_model_id = Column(String(255))  # For Bedrock
    aws_sagemaker_endpoint_name = Column(String(255))  # For SageMaker
    aws_content_handler_class = Column(String(255))  # For SageMaker
    
    # Google Cloud Vertex AI fields
    gcp_project_id = Column(String(255))
    gcp_region = Column(String(50))
    gcp_auth_method = Column(String(50))  # adc, service_account_key
    gcp_service_account_key = Column(Text)  # Encrypted JSON
    gcp_model_type = Column(String(50))  # foundation, custom_endpoint
    gcp_model_name = Column(String(255))  # For foundation models
    gcp_endpoint_id = Column(String(255))  # For custom endpoints
    gcp_model_id = Column(String(255))  # For custom endpoints
    
    # Custom Deployment fields
    custom_deployment_location = Column(String(100))  # on_premise, azure_compute, aws_compute, gcp_compute, other_cloud
    custom_llm_provider = Column(String(255))
    custom_api_endpoint_url = Column(String(500))
    custom_api_compatibility = Column(String(50))  # openai_compatible, huggingface_tgi, ollama, custom
    custom_auth_method = Column(String(50))  # none, api_key_header, bearer_token, other
    custom_auth_header_name = Column(String(100))
    custom_auth_key_prefix = Column(String(50))
    custom_auth_api_key = Column(Text)  # Encrypted
    
    # General LLM Configuration
    model_name = Column(String(255))  # Internal tracking name
    temperature = Column(Float)
    max_tokens = Column(Integer)
    top_p = Column(Float)
    stop_sequences = Column(Text)  # Comma-separated
    
    # System fields
    enabled = Column(Boolean, default=True)
    status = Column(String(50), default="active")
    usage_stats = Column(JSON)
    additional_config = Column(JSON)  # For any extra configuration
